File: tools/lib/zeo_dev/commands/corelib.rb
# frozen_string_literal: true

require "digest"
require "fileutils"
require "json"

module ZeoDev
  module Commands
    # The Ruby files CRuby compiles INTO the interpreter, vendored verbatim.
    #
    #   corelib verify          re-hash the committed files against the lock
    #   corelib sync [--check]  re-vendor them from the pinned rev
    #   corelib census          which upstream corelib files are eligible
    #
    # Two checks with different reach, and the split is the point.
    #
    # `verify` is OFFLINE. It re-hashes `crates/zeo/corelib/*.rb` and compares
    # to `upstream.lock`, so it runs in CI, in a release tarball, and on a
    # machine with no network -- and it is the only one a downstream consumer
    # can run. It proves the committed bytes are the bytes that were locked.
    #
    # `sync --check` needs the NETWORK. It fetches `ruby/ruby` at the pinned
    # rev and compares the upstream bytes to the committed ones, which is what
    # proves the locked bytes are upstream's. Run it when the pin moves.
    #
    # The lock also records each file's git blob OID -- `sha1("blob <n>\0" +
    # bytes)`, which is what GitHub's contents API answers -- so a reader can
    # verify a file against github.com with one request and no clone:
    #
    #   curl -s https://api.github.com/repos/ruby/ruby/contents/nilclass.rb\
    #     ?ref=<rev> | jq -r .sha
    class Corelib < Cli
      SUBCOMMANDS = %w[verify sync census].freeze

      def self.summary = "verify or re-vendor the corelib ruby"

      def self.banner
        <<~USAGE
          usage: zeo-dev corelib <subcommand>

            verify          re-hash the committed files against upstream.lock (offline)
            sync [--check]  re-vendor from the pinned rev (network)
            census          list which upstream corelib files are eligible
        USAGE
      end

      def defaults = { check: false }

      def options(o)
        o.on("--check", "verify without writing; nonzero on drift") { opts[:check] = true }
      end

      def run
        sub = args.shift
        raise Error, self.class.banner unless SUBCOMMANDS.include?(sub)

        send(:"cmd_#{sub}")
      end

      private

      def manifest = @manifest ||= Manifest.load(ROOT)

      def entry
        manifest.group(:corelib).first or raise Error, "upstream.rb declares no corelib entry"
      end

      def dir = File.join(ROOT, Manifest::CORELIB_DIR)

      # Offline: the committed bytes against the lock's digests.
      def cmd_verify
        locked = JSON.parse(File.read(manifest.lock_path))["corelib"] || []
        rows = locked.flat_map { |e| e["files"] || [] }
        if rows.empty?
          warn "corelib: upstream.lock records no corelib files"
          return 1
        end
        bad = 0
        # A name in `upstream.rb` with no row here was never vendored -- see
        # `Manifest.corelib_digests`. Caught before the hashes, because a
        # missing file is a missing check rather than a failing one.
        (entry.files.sort - rows.map { |r| r["path"] }).each do |rel|
          warn "corelib: #{rel} is named in upstream.rb but not vendored -- run `corelib sync`"
          bad += 1
        end
        rows.each do |row|
          path = File.join(dir, row["path"])
          unless File.file?(path)
            warn "corelib: #{row["path"]} is MISSING"
            bad += 1
            next
          end
          bytes = File.binread(path)
          sha = Digest::SHA256.hexdigest(bytes)
          blob = Manifest.blob_oid(bytes)
          if sha == row["sha256"] && blob == row["blob"]
            puts "corelib: #{row["path"]} ok (blob #{blob[0, 12]})"
          else
            warn "corelib: #{row["path"]} DOES NOT MATCH upstream.lock"
            warn "  locked sha256 #{row["sha256"]}"
            warn "  actual sha256 #{sha}"
            bad += 1
          end
        end
        return 1 if bad.positive?

        puts "corelib: #{rows.size} file(s) match #{entry.github} @ #{entry.tag} (#{entry.short_rev})"
        0
      end

      # Network: upstream at the pinned rev against the committed bytes.
      def cmd_sync
        checkout = Vendor.fetch_checkout(entry, entry.rev)
        drift = 0
        entry.files.sort.each do |rel|
          upstream = File.join(Vendor.source_root(checkout, entry), rel)
          raise Error, "#{entry.github} @ #{entry.tag} has no #{rel}" unless File.file?(upstream)

          bytes = File.binread(upstream)
          reject_primitives(rel, bytes)
          dest = File.join(dir, rel)
          if opts[:check]
            same = File.file?(dest) && File.binread(dest) == bytes
            puts(same ? "corelib: #{rel} is upstream #{entry.tag}" : "corelib: #{rel} DIFFERS from upstream #{entry.tag}")
            drift += 1 unless same
          else
            FileUtils.mkdir_p(File.dirname(dest))
            File.binwrite(dest, bytes)
            puts "corelib: vendored #{rel} (#{bytes.bytesize} bytes)"
          end
        end
        return drift.zero? ? 0 : 1 if opts[:check]

        # Re-loaded, not reused: the digests are derived from the bytes on
        # disk, and the bytes just changed.
        Manifest.load(ROOT).write_lock!
        puts "corelib: re-locked #{Manifest::LOCK}"
        0
      end

      # A corelib file that reaches for a primitive is not vendorable: the
      # `Primitive.`/`__builtin`/`cexpr!` protocol is a C-level escape hatch
      # zeo has no answer for, and a file carrying one would compile and then
      # fail at the call. Refusing here is what keeps `corelib/` honest.
      def reject_primitives(rel, bytes)
        found = %w[Primitive. __builtin cexpr! cstmt! attr!].select { |m| bytes.include?(m) }
        return if found.empty?

        raise Error, "#{rel} uses #{found.join(", ")} -- the primitive protocol is not implemented, " \
                     "so this file cannot be corelib (see upstream.rb)"
      end

      # Which of upstream's `BUILTIN_RB_SRCS` carry no primitive at all.
      def cmd_census
        checkout = Vendor.fetch_checkout(entry, entry.rev)
        mk = File.read(File.join(checkout, "common.mk"))
        list = mk[/^BUILTIN_RB_SRCS = \\\n(.*?)^\s*\$\(empty\)/m].to_s
        names = list.scan(%r{\$\(srcdir\)/(\S+\.rb)}).flatten.sort
        vendored = entry.files.sort
        names.each do |rel|
          path = File.join(checkout, rel)
          next unless File.file?(path)

          bytes = File.binread(path)
          marks = %w[Primitive. __builtin cexpr! cstmt! attr!].sum { |m| bytes.scan(m).size }
          state = if vendored.include?(rel) then "vendored"
                  elsif marks.zero? then "ELIGIBLE"
                  else "needs #{marks} primitive(s)"
                  end
          puts format("%-24s %6d lines  %s", rel, bytes.count("\n"), state)
        end
        0
      end
    end
  end
end

File: crates/zeo/src/parse/corelib.rs
//! The Ruby files CRuby compiles INTO its own interpreter, compiled into a
//! zeo program the same way.
//!
//! CRuby writes 24 of its core files in Ruby (`BUILTIN_RB_SRCS` in
//! `common.mk`) and compiles them into the binary; a method defined in one
//! reports `<internal:nilclass>` as its source. zeo vendors those files
//! VERBATIM from the same `ruby/ruby` pin the C API headers ride, so the rows
//! a program dispatches to are CRuby's own code rather than a Rust
//! reimplementation of it -- which is the only way the signatures, the
//! `source_location`s and the corner cases can agree by construction.
//!
//! # What is eligible
//!
//! Only a file with NO `Primitive.`, `__builtin` or `cexpr!` in it. Those are
//! a C-level escape hatch zeo has no answer for, and a file carrying one
//! would compile and then fail at the call. `tools/zeo-dev corelib census`
//! prints the eligibility of all 24; at v4.0.6 five qualify and this module
//! carries the ones zeo has measured.
//!
//! # Provenance
//!
//! `crates/zeo/corelib/` is upstream's bytes and carries NO patch series --
//! unlike the C headers next door, which describe zeo's object layout rather
//! than ruby's behaviour. Three checks hold the claim up, and each covers what
//! the others cannot:
//!
//! * `build.rs` hashes each file against `upstream.lock` and stages it into
//!   `OUT_DIR`. The `include_str!`s below read the STAGED copy, so bytes that
//!   were not hashed cannot reach a compiled program.
//! * `tools/zeo-dev corelib verify` re-hashes offline -- in CI, in a release
//!   tarball, on a machine with no network.
//! * `tools/zeo-dev corelib sync --check` fetches the pinned rev and proves
//!   the locked bytes ARE upstream's.
//!
//! The lock also records each file's git blob OID, which is what GitHub's
//! contents API answers, so a reader can verify one against github.com with a
//! single request and no clone.

use crate::hir::{Hir, NodeId};
use crate::lower_error::LowerError;

/// One vendored file: what it is called, what CRuby reports for it, and
/// whether it is compiled in unless asked otherwise.
struct Segment {
    /// The name `ZEO_CORELIB` selects it by.
    name: &'static str,
    /// What `#source_location` and a frame answer. CRuby's own spelling --
    /// `<internal:nilclass>`, not a path -- because that is what a program
    /// reading either one gets from ruby.
    internal_name: &'static str,
    source: &'static str,
    /// Whether it is on with no `ZEO_CORELIB` set.
    ///
    /// This is a SIZE decision, and the only one. A Ruby row is emitted into
    /// every binary that can reach it, while a Rust row already lives in the
    /// shared archive -- so `nilclass` costs 34 KB and is on, and `pathname`
    /// costs 8.7 MB and is not. Nothing about correctness differs: both are
    /// CRuby's own bytes and both pass the corpus.
    ///
    /// Gating `pathname` on `require "pathname"` is NOT available, and that
    /// is a fact about ruby rather than a limitation here: ruby 4.0 loads
    /// `pathname.so` before the first line, so `Pathname.new` answers with no
    /// require at all and only `#find`/`#rmtree` arrive with one.
    on_by_default: bool,
}

impl Segment {
    /// Whether this compile carries the segment.
    fn selected(&self, policy: &crate::Corelib) -> bool {
        match policy {
            crate::Corelib::Default => self.on_by_default,
            crate::Corelib::Rust => false,
            crate::Corelib::Only(names) => names.iter().any(|n| n == self.name),
        }
    }
}

/// The vendored corelib, in the order it is compiled in.
const SEGMENTS: &[Segment] = &[
    Segment {
        name: "nilclass",
        internal_name: "<internal:nilclass>",
        source: include_str!(concat!(env!("OUT_DIR"), "/corelib_nilclass.rb")),
        on_by_default: true,
    },
    Segment {
        name: "pathname",
        internal_name: "<internal:pathname_builtin>",
        source: include_str!(concat!(env!("OUT_DIR"), "/corelib_pathname_builtin.rb")),
        on_by_default: false,
    },
];

/// Lower the SELECTED segments into `hir`, appending their statements.
///
/// Per compile, not once: the exception prelude is cached in a template arena
/// and cloned, and an unselected corelib segment must not be lowered into it
/// at all. Leaving one there as dead nodes is not free -- `uses_runtime_eval`
/// and every other whole-arena scan reads `Hir::nodes` directly, so a
/// segment nobody selected still linked the compiler into every binary.
///
/// Each file is registered in the source table under its `<internal:>` name
/// and lowered with `lowering_file` pointing at it, so a `def` inside one
/// carries that name and the file's own line -- which is what makes
/// `NilClass.instance_method(:to_i).source_location` answer
/// `["<internal:nilclass>", 36]` the way ruby's does.
pub(super) fn lower_into(
    hir: &mut Hir,
    statements: &mut Vec<NodeId>,
    policy: &crate::Corelib,
) -> Result<(), LowerError> {
    for seg in SEGMENTS.iter().filter(|s| s.selected(policy)) {
        lower_as_internal(hir, statements, seg.internal_name, seg.source)?;
    }
    Ok(())
}

/// Lower `source` as an INTERNAL file named `internal_name`, appending its
/// statements.
///
/// One place registers the file and marks it internal, so every carve-out
/// keyed on `Hir::internal_files` -- the `method_added` hook, `Coverage`, and
/// the fused-iterator guard -- covers a segment by construction rather than
/// by each remembering to.
pub(crate) fn lower_as_internal(
    hir: &mut Hir,
    statements: &mut Vec<NodeId>,
    internal_name: &str,
    source: &'static str,
) -> Result<(), LowerError> {
    let file = hir.add_file(internal_name, source);
    // Not part of the program in any sense a program can observe --
    // see `Hir::internal_files`.
    hir.internal_files.insert(file);
    let saved = hir.lowering_file.replace(file);
    let lowered = crate::lower::parse_and_lower_into(hir, source).map_err(|e| LowerError {
        message: format!(
            "internal error in zeo's vendored corelib {internal_name} (this is a zeo bug): {}",
            e.message
        ),
        ..e
    });
    hir.lowering_file = saved;
    statements.extend(lowered?);
    Ok(())
}

/// Every segment name, for the CLI's error when `ZEO_CORELIB` names one that
/// does not exist -- a typo there is otherwise a silent no-op.
pub(crate) fn names() -> Vec<&'static str> {
    SEGMENTS.iter().map(|s| s.name).collect()
}
